main.go

```diff
@@ -159,6 +270,45 @@ func newController(klient kubernetes.Interface, logger log.Logger, o *options) *
 		ssetInf: appsinformers.NewFilteredStatefulSetInformer(klient, o.namespace, resyncPeriod, nil, func(lo *v1.ListOptions) {
 			lo.LabelSelector = labels.Set{o.labelKey: o.labelValue}.String()
 		}),
+
+		reconcileAttempts: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "thanos_receive_controller_reconcile_attempts_total",
+			Help: "Total number of reconciles.",
+		}),
+
+		reconcileErrors: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: "thanos_receive_controller_reconcile_errors_total",
+			Help: "Total number of reconciles errors.",
+		},
+			[]string{"type"},
+		),
+
+		configmapChangeAttempts: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: "thanos_receive_controller_configmap_change_attempts_total",
+			Help: "Total number of configmap change attempts.",
+		},
+			[]string{"verb"},
+		),
+
+		configmapChangeErrors: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "thanos_receive_controller_configmap_change_errors_total",
+			Help: "Total number of configmap change errors.",
+		}),
+	}
+}
+
+func (c *controller) registerMetrics(reg *prometheus.Registry) {
+	if reg != nil {
+		c.reconcileAttempts.Add(0)
+		c.reconcileErrors.WithLabelValues().Add(0)
```

Comment:
either we initialize this for all the values we know there will be or not at all